# Data Cleaning

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime, date
import warnings

warnings.filterwarnings("ignore")

To begin, I will first load the data of each month into a separate data frame and later on merge them all into two separate frames (one for 2019 and one for 2020) for easier cleaning. 

Since March 8, 2020 is widely viewed as the start of the pandemic -- this is when Italy first went into lockdown-- I will filter March's dataframe to reflect the data starting from the 8th.

In [2]:
march = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\march_2020.csv')
march = march.loc[(pd.DatetimeIndex(march["tpep_pickup_datetime"]).month == 3) & 
              (pd.DatetimeIndex(march["tpep_pickup_datetime"]).day >=8)]

**Note:** As seen above, according to our data dictionary, **tpep_pickup_datetime** means the date and time when the [taxi] meter was engaged.

A quick check:

In [3]:
march.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1497226       2.0  2020-03-08 00:00:33   2020-03-08 00:11:14              2.0   
1497353       2.0  2020-03-08 00:02:10   2020-03-08 00:06:24              1.0   
1497609       2.0  2020-03-08 00:01:34   2020-03-08 00:04:34              1.0   
1498278       2.0  2020-03-08 00:00:21   2020-03-08 00:25:46              2.0   
1498328       2.0  2020-03-08 00:00:10   2020-03-08 00:06:23              2.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1497226           1.06         1.0                  N           148   
1497353           0.73         1.0                  N            74   
1497609           0.48         1.0                  N           236   
1498278          18.62         2.0                  N           132   
1498328           1.30         1.0                  N           239   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1497226             4           1.0          8.0    0.5      0.5        2.36   
1497353            41           2.0          5.0    0.5      0.5        0.00   
1497609           262           1.0          4.0    0.5      0.5        1.00   
1498278            42           1.0         52.0    0.0      0.5       11.78   
1498328           163           1.0          7.0    0.5      0.5        2.16   

         tolls_amount  improvement_surcharge  total_amount  \
1497226          0.00                    0.3         14.16   
1497353          0.00                    0.3          6.30   
1497609          0.00                    0.3          8.80   
1498278          6.12                    0.3         70.70   
1498328          0.00                    0.3         12.96   

         congestion_surcharge  
1497226                   2.5  
1497353                   0.0  
1497609                   2.5  
1498278                   0.0  
1498328                   2.5

In [4]:
april = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\april_2020.csv')
may = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\may_2020.csv')
june = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\june_2020.csv')
july = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\july_2020.csv')
august = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\august_2020.csv')
september = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\september_2020.csv')
october = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\october_2020.csv')
november = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\november_2020.csv')
december = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\december_2020.csv')

In [5]:
year_20= pd.concat([march, april, may, june, july, august, september, october, november, december])

A quick check of our full dataframe:

In [6]:
year_20.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1497226       2.0  2020-03-08 00:00:33   2020-03-08 00:11:14              2.0   
1497353       2.0  2020-03-08 00:02:10   2020-03-08 00:06:24              1.0   
1497609       2.0  2020-03-08 00:01:34   2020-03-08 00:04:34              1.0   
1498278       2.0  2020-03-08 00:00:21   2020-03-08 00:25:46              2.0   
1498328       2.0  2020-03-08 00:00:10   2020-03-08 00:06:23              2.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1497226           1.06         1.0                  N           148   
1497353           0.73         1.0                  N            74   
1497609           0.48         1.0                  N           236   
1498278          18.62         2.0                  N           132   
1498328           1.30         1.0                  N           239   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1497226             4           1.0          8.0    0.5      0.5        2.36   
1497353            41           2.0          5.0    0.5      0.5        0.00   
1497609           262           1.0          4.0    0.5      0.5        1.00   
1498278            42           1.0         52.0    0.0      0.5       11.78   
1498328           163           1.0          7.0    0.5      0.5        2.16   

         tolls_amount  improvement_surcharge  total_amount  \
1497226          0.00                    0.3         14.16   
1497353          0.00                    0.3          6.30   
1497609          0.00                    0.3          8.80   
1498278          6.12                    0.3         70.70   
1498328          0.00                    0.3         12.96   

         congestion_surcharge  
1497226                   2.5  
1497353                   0.0  
1497609                   2.5  
1498278                   0.0  
1498328                   2.5

In [7]:
len(year_20)

10421450

Now I will check for the percentages of null values, if any, and decide how to handle accordingly.

In [8]:
null_values = year_20.isna().sum()
null_percentage = null_values[null_values>0] / len(year_20)
null_percentage.to_frame('% Null')

% Null
VendorID            0.06547
passenger_count     0.06547
RatecodeID          0.06547
store_and_fwd_flag  0.06547
payment_type        0.06547

Since the null values are very low (much less than half a percentage point), I've decided to remove all the null values to avoid the possibilty of being thrown an error (should I leave them), or having incorrect results if I were to replace them (with a mean value for example). I've done so because the **payment_type** column is vital to a question I want to answer, so I want to produce as accurate results as possible in my analysis.

I will drop all of the NaN/Null values and perform a quick check of the beginning and end of the dataframe:

In [9]:
year_20.dropna(inplace=True)
year_20.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1497226       2.0  2020-03-08 00:00:33   2020-03-08 00:11:14              2.0   
1497353       2.0  2020-03-08 00:02:10   2020-03-08 00:06:24              1.0   
1497609       2.0  2020-03-08 00:01:34   2020-03-08 00:04:34              1.0   
1498278       2.0  2020-03-08 00:00:21   2020-03-08 00:25:46              2.0   
1498328       2.0  2020-03-08 00:00:10   2020-03-08 00:06:23              2.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1497226           1.06         1.0                  N           148   
1497353           0.73         1.0                  N            74   
1497609           0.48         1.0                  N           236   
1498278          18.62         2.0                  N           132   
1498328           1.30         1.0                  N           239   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1497226             4           1.0          8.0    0.5      0.5        2.36   
1497353            41           2.0          5.0    0.5      0.5        0.00   
1497609           262           1.0          4.0    0.5      0.5        1.00   
1498278            42           1.0         52.0    0.0      0.5       11.78   
1498328           163           1.0          7.0    0.5      0.5        2.16   

         tolls_amount  improvement_surcharge  total_amount  \
1497226          0.00                    0.3         14.16   
1497353          0.00                    0.3          6.30   
1497609          0.00                    0.3          8.80   
1498278          6.12                    0.3         70.70   
1498328          0.00                    0.3         12.96   

         congestion_surcharge  
1497226                   2.5  
1497353                   0.0  
1497609                   2.5  
1498278                   0.0  
1498328                   2.5

In [38]:
year_20.tail()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1362436       2.0  2020-12-31 23:05:33   2020-12-31 23:31:36              1.0   
1362437       2.0  2020-12-31 22:57:20   2020-12-31 23:05:33              1.0   
1362438       2.0  2020-12-31 23:40:35   2020-12-31 23:48:43              1.0   
1362439       2.0  2020-12-31 23:54:57   2020-12-31 23:57:39              1.0   
1362440       2.0  2020-12-31 23:11:16   2020-12-31 23:24:08              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1362436          11.30         1.0                  N           107   
1362437           2.18         1.0                  N           236   
1362438           2.52         1.0                  N           236   
1362439           0.59         1.0                  N           238   
1362440           6.06         1.0                  N            75   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1362436           242           2.0         33.0    0.5      0.5        0.00   
1362437           161           1.0          9.0    0.5      0.5        2.56   
1362438            24           1.0          9.5    0.5      0.5        4.00   
1362439           239           1.0          4.5    0.5      0.5        2.08   
1362440           169           2.0         18.5    0.5      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
1362436           0.0                    0.3         36.80   
1362437           0.0                    0.3         15.36   
1362438           0.0                    0.3         17.30   
1362439           0.0                    0.3         10.38   
1362440           0.0                    0.3         19.80   

         congestion_surcharge  
1362436                   2.5  
1362437                   2.5  
1362438                   2.5  
1362439                   2.5  
1362440                   0.0

A quick check to be sure all null values were dropped and to see how many rows were removed:

In [10]:
null_values = year_20.isna().sum()
null_percentage = null_values[null_values>0] / len(year_20)
null_percentage.to_frame('% Null')

Empty DataFrame
Columns: [% Null]
Index: []

In [11]:
len(year_20)

9739158

In [12]:
10421450 - 9739158

682292

So, out of over 10 million entries, only 68,000 contained null values and were subsequently removed.

Now, I will repeat the above steps for the data from 2019, or pre-covid.

In [13]:
march = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\march_2019.csv')
#To achieve a balanced analysis, I will filter the 2019 March data to reflect the same time period as 2020's (starting
#from the 8th)
march = march.loc[(pd.DatetimeIndex(march["tpep_pickup_datetime"]).month == 3) & 
              (pd.DatetimeIndex(march["tpep_pickup_datetime"]).day >=8)]
april = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\april_2019.csv')
may = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\may_2019.csv')
june = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\june_2019.csv')
july = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\july_2019.csv')
august = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\august_2019.csv')
september = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\september_2019.csv')
october = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\october_2019.csv')
november = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\november_2019.csv')
december = pd.read_csv('C:\\Users\\15164\\Desktop\\nyc-taxis-vs-covid\\data\\december_2019.csv')

In [14]:
year_19= pd.concat([march, april, may, june, july, august, september, october, november, december])

In [15]:
year_19.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
130887        2.0  2019-03-22 08:22:39   2019-03-22 08:32:50              1.0   
160292        2.0  2019-03-21 15:31:46   2019-03-21 15:55:19              1.0   
160293        2.0  2019-03-21 16:03:03   2019-03-21 16:15:59              1.0   
160294        2.0  2019-03-21 16:18:06   2019-03-21 23:27:41              1.0   
1612737       2.0  2019-03-08 00:15:18   2019-03-08 00:22:43              6.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
130887            1.45         1.0                  N           170   
160292            2.00         1.0                  N           186   
160293            1.36         1.0                  N           161   
160294            0.97         1.0                  N           141   
1612737           1.25         1.0                  N           249   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
130887            107           1.0          8.5    0.0      0.5        2.95   
160292            163           1.0         15.0    1.0      0.5        1.50   
160293            237           1.0          9.5    1.0      0.5        2.76   
160294            236           1.0          7.5    1.0      0.5        1.50   
1612737            79           1.0          7.0    0.0      0.5        1.00   

         tolls_amount  improvement_surcharge  total_amount  \
130887            0.0                    0.3         14.75   
160292            0.0                    0.3         20.80   
160293            0.0                    0.3         16.56   
160294            0.0                    0.3         13.30   
1612737           0.0                    0.3         11.30   

         congestion_surcharge  
130887                    2.5  
160292                    2.5  
160293                    2.5  
160294                    2.5  
1612737                   2.5

In [16]:
len(year_19)

67904737

*Note: As we can already see, there is a **significant** difference in the amount of taxi trips between 2019 and 2020!*

In [17]:
null_values = year_19.isna().sum()
null_percentage = null_values[null_values>0] / len(year_19)
null_percentage.to_frame('% Null')

% Null
VendorID              3.631573e-03
passenger_count       3.631573e-03
RatecodeID            3.631573e-03
store_and_fwd_flag    3.631573e-03
payment_type          3.631573e-03
congestion_surcharge  4.417954e-08

In [18]:
year_19.dropna(inplace=True)
len(year_19)

67658133

In [20]:
print(67904737 - 67658133)
print((67904737 - 67658133) / 67904737 * 100)

246604
0.36316170402073716


After removing the null values, I've dropped 246,604 rows or 0.36% of the data.

In [21]:
year_19.to_csv('taxi_2019.csv')

In [22]:
year_20.to_csv('taxi_2020.csv')